rss2email/feeds.py

```diff
@@ -337,32 +339,37 @@ def save(self):
         dst_config_file = _os.path.realpath(self.configfiles[-1])
         _LOG.debug('save feed configuration to {}'.format(dst_config_file))
         for feed in self:
+            # Saves feed config to the config object, not the actual file (yet)
             feed.save_to_config()
         dirname = _os.path.dirname(dst_config_file)
         if dirname and not _os.path.isdir(dirname):
             _os.makedirs(dirname, mode=0o700, exist_ok=True)
-        tmpfile = dst_config_file + '.tmp'
-        with open(tmpfile, 'w') as f:
+
+        with open(dst_config_file, 'w') as f:
+            _fcntl.lockf(f, _fcntl.LOCK_EX) # lock for writing
             self.config.write(f)
             f.flush()
             _os.fsync(f.fileno())
-        _os.replace(tmpfile, dst_config_file)
+
         self._save_feeds()
 
     def _save_feeds(self):
         _LOG.debug('save feed data to {}'.format(self.datafile_path))
         dirname = _os.path.dirname(self.datafile_path)
         if dirname and not _os.path.isdir(dirname):
             _os.makedirs(dirname, mode=0o700, exist_ok=True)
-        tmpfile = self.datafile_path + '.tmp'
-        with _codecs.open(tmpfile, 'w', self.datafile_encoding) as f:
-            self._save_feed_states(feeds=self, stream=f)
-            f.flush()
-            _os.fsync(f.fileno())
-        _os.replace(tmpfile, self.datafile_path)
-        if self._datafile is not None:
-            self._datafile.close()  # release the lock
-            self._datafile = None
+
+        # Note: the data file is locked for writing for the entire lifetime
+        # of the feeds object, so the following is safe.
+
+        # We don't want to close the file and release the lock (allowing
+        # another rss2email to overwrite the data under us), so we do a
+        # seek and truncate
+        self.datafile.seek(0)
+        self._save_feed_states(feeds=self, stream=self.datafile)
+        self.datafile.truncate()
+        self.datafile.flush()
+        _os.fsync(self.datafile.fileno())
```

Comment:
Suggestion: for the same reason as for the config, either add a `# TODO figure out a way to make this write atomic` or figure the thing out now, by doing something like write to a tmp file, locking the tmp file, mv-ing it in place and then closing the now-deleted file -- I'm not sure that's proper posix, though.

Reply:
Yeah, if r2e dies in the middle, bad things will happen, what I am doing is not at all atomic. The only way out, I suppose, is to have a dedicated lock file which we acquire an exclusive lock on right at the start, then we do the atomic copy to temp file, modify, and rename.